File: frontend/src/navigation.ts
import { Query } from "@/api/types";

type PostSearchParams = Query<"/api/posts/">;

const Navigation = {
  home: () => "/",
  contact: () => "/contact/",
  about: (path?: string) => (path ? `/about/${path}` : "/about/"),
  app: (slug: string) => `/apps/${slug}/`,
  appLiveInstance: (slug: string) => `/apps/${slug}/live/`,
  changelog: (slug: string) => `/changelog/${slug}/`,
  post: (slug: string) => `/posts/${slug}/`,
  posts: (params?: PostSearchParams) => `/${searchParams(params)}`,
  feed: (params?: PostSearchParams) => `/feed/${searchParams(params)}`,
  tag: (tag: string) => Navigation.posts({ tag }),
  webmentionsTest: () => `/webmentions_tester/`,
};
type Navigable = keyof typeof Navigation;
export type NavDestination = {
  [K in Navigable]: (typeof Navigation)[K] extends () => string ? K : never;
}[Navigable];

export const navigationHref = <T extends Navigable>(
  type: T,
  ...args: Parameters<(typeof Navigation)[T]>
): string => {
  type P = (...args: Parameters<(typeof Navigation)[T]>) => string;
  return (Navigation[type] as P)(...args);
};

const searchParams = (params: Record<string, string | number> | undefined) => {
  if (!params) return "";

  const formatted = Object.entries(params)
    .filter(([key, value]) => !!value)
    .map(([key, value]) => `${key}=${value}`)
    .join("&");
  if (!formatted) return "";
  return `?${formatted}`;
};

export const absoluteUrl = (path: string | undefined): string | undefined =>
  path ? `${process.env.NEXT_PUBLIC_SITE_BASE_URL}${path}` : undefined;

File: beatonma-django/navigation.py
from common.util.url import enforce_trailing_slash


def home():
    return "/"


def contact():
    return "/contact/"


def about(path: str | None = None):
    return enforce_trailing_slash(f"/about/{path}")


def app(slug: str):
    return f"/apps/{slug}/"


def app_live_instance(slug: str):
    return f"/apps/{slug}/live/"


def changelog(slug: str):
    return f"/changelog/{slug}/"


def post(slug: str):
    return f"/posts/{slug}/"


def posts(*, tag: str = None, query: str = None):
    params = {"tag": tag, "query": query}
    params = [f"{k}={v}" for k, v in params.items() if v]
    params = "&".join(params)
    return "?".join([f"/", params])


def tag(tag: str):
    return posts(tag=tag.removeprefix("#"))


def webmentions_test():
    return "/webmentions_tester/"
